### ISSUE TITLE
Parallel compilation is broken with GNU Make 4.4

### ISSUE THREAD

**Issue by mukilan** [servo/mozjs#375]
GNU Make 4.4 breaks the parallel builds in mozjs and behaves as though `-j1` was given, leading to long build times.

Relevant logs from `cargo build -vv`:
```
[mozjs_sys 0.68.2] make -f Makefile
[mozjs_sys 0.68.2] make[1]: Entering directory '/home/mukilan/dev/mozjs/target/debug/build/mozjs_sys-094547916dc1fb70/out/build'
[mozjs_sys 0.68.2] make[2]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
[mozjs_sys 0.68.2] make[2]: Entering directory '/home/mukilan/dev/mozjs/target/debug/build/mozjs_sys-094547916dc1fb70/out/build'
[mozjs_sys 0.68.2] make recurse_pre-export
```

Th release notes for [GNU Make 4.4](https://lists.gnu.org/archive/html/help-make/2022-10/msg00020.html) and the linked bugs note that when using 'old-style' anonymous pipes for jobserver, the file descriptors are closed when parent make invokes a recursive sub-make when the sub-make call is *[not direct](https://www.gnu.org/software/make/manual/make.html#MAKE-Variable)* 
```
* When the pipe-based jobserver is enabled and GNU Make decides it is invoking
  a non-make sub-process and closes the jobserver pipes, it will now add a new
  option to the MAKEFLAGS environment variable that disables the jobserver.
  This prevents sub-processes that invoke make from accidentally using other
  open file descriptors as jobserver pipes.  For more information see
  https://savannah.gnu.org/bugs/?57242 and https://savannah.gnu.org/bugs/?62397
```

SpiderMonkey's Makefile seem to implement recursive invocations that are [not direct](https://searchfox.org/mozilla-central/source/config/rules.mk#233) $(MAKE) invocations, so these sub-makes fail to open the closed jobserver pipes and fallback to -j1 mode.

This issue can be fixed by using a version of [Cargo that has support](https://github.com/rust-lang/cargo/blob/master/CHANGELOG.md#cargo-169-2023-04-20) for the new 'named fifo' based jobserver as GNU Make 4.4 by introduced them to allow  any process to be integrate with the jobserver and thus should work with both direct and indirect sub-make invocations.

This affects Servo (on distros that ship GNU Make 4.4 such as Fedora 39, Arch,  NixOS  23.05 etc) since it currently uses 2023-02-01 build of rust toolchain where the Cargo doesn't have support for named fifo jobserver.

**Comment by sagudev** [comment]
I think the best way would be to simply update servo's rust version. There is already PR for that: https://github.com/servo/servo/pull/29993

**Comment by mukilan** [comment]
Yep, I've called that out in the [PR in servo](https://github.com/servo/servo/pull/30146) to pin GNU Make :)

**Comment by delan** [comment]
It seems moving to Rust 1.74 was not enough to get make 4.4 working well :(

When building mozjs_sys as part of <https://github.com/servo/servo/commit/355567186400d0531657d0eedda7625de9d0e93a> with nixpkgs pinned to nixos-23.05 (<https://github.com/NixOS/nixpkgs/commit/70bdadeb94ffc8806c0570eb5c2695ad29f0e421>), the build script takes 634 seconds ([cargo-timing-20240110T071341Z.html.gz](https://github.com/servo/mozjs/files/13884552/cargo-timing-20240110T071341Z.html.gz)).

When gnumake is downgraded from 4.4.1 to 4.3 (<https://github.com/NixOS/nixpkgs/commit/6adf48f53d819a7b6e15672817fa1e78e5f4e84f>), the build script takes only 71 seconds on the same 16-core 32-thread machine ([cargo-timing-20240110T074057Z.html.gz](https://github.com/servo/mozjs/files/13884594/cargo-timing-20240110T074057Z.html.gz)).

```diff
diff --git a/etc/shell.nix b/etc/shell.nix
index 306b25acb9..81deee77fe 100644
--- a/etc/shell.nix
+++ b/etc/shell.nix
@@ -2,7 +2,7 @@
 # NOTE: This does not work offline or for nix-build
 
 with import (builtins.fetchTarball {
-  url = "https://github.com/NixOS/nixpkgs/archive/6adf48f53d819a7b6e15672817fa1e78e5f4e84f.tar.gz";
+  url = "https://github.com/NixOS/nixpkgs/archive/70bdadeb94ffc8806c0570eb5c2695ad29f0e421.tar.gz";
 }) {
   overlays = [
     (import (builtins.fetchTarball {
@@ -17,6 +17,9 @@ let
       cargo = rustToolchain;
       rustc = rustToolchain;
     };
+    pkgs_gnumake_4_3 = import (builtins.fetchTarball {
+      url = "https://github.com/NixOS/nixpkgs/archive/6adf48f53d819a7b6e15672817fa1e78e5f4e84f.tar.gz";
+    }) {};
 in
 clangStdenv.mkDerivation rec {
   name = "servo-env";
@@ -42,7 +45,7 @@ clangStdenv.mkDerivation rec {
     # functionality in mozjs and causes builds to be extremely
     # slow as it behaves as if -j1 was passed.
     # See https://github.com/servo/mozjs/issues/375
-    gnumake
+    pkgs_gnumake_4_3.gnumake
 
     # crown needs to be in our Cargo workspace so we can test it with `mach test`. This means its
     # dependency tree is listed in the main Cargo.lock, making it awkward to build with Nix because
```

```
$ nice ./mach build -d
$ rm -Rf target/debug/{build,deps,incremental}/mozjs*
$ find target/ -samefile target/debug/servo -print -delete
$ nice ./mach build -d
```

**Comment by sagudev** [comment]
Could this be problematic? https://github.com/servo/mozjs/blob/7184f658b7e2de8a8d288ff39001364513115782/mozjs-sys/makefile.cargo#L12

**Comment by mukilan** [comment]
I printed the value of `CARGO_MAKEFLAGS` when building mozjsys and it looks like it still uses pipes

```
  "-j --jobserver-fds=7,8 --jobserver-auth=7,8"
```

It looks like cargo for now only supports *inheriting and forwarding* named fifo arguments and new jobservers that cargo itself creates are still using pipes:

From https://docs.rs/jobserver/0.1.27/jobserver/ 

> Starting from GNU make version 4.4, named pipe becomes the default way in communication on Unix. This crate also supports that feature in the sense of inheriting and forwarding the correct environment.

And https://github.com/rust-lang/jobserver-rs/pull/49

> This commit makes sure that the new style --jobserver-auth=fifo:PATH
> can be forwarded to inherited processes correctly.
> The support of creating a new client with named pipe will come as follow-up pull request.

**Comment by mukilan** [comment]
> Can the upstream Makefile be fixed

I tried a couple of approaches to fix the makefile back when I was investigating this issue, but I didn't find any working solutions. Admittedly my knowledge of Make and SM build system is also limited, so in theory a fix to the Makefile might be still be possible. 

> Can we run a fifo-based jobserver [with jobslot](https://docs.rs/jobslot/latest/jobslot/struct.Client.html#method.new_with_fifo)?

That's an interesting idea! I wasn't aware of jobslot. For the whole Servo build to use the same job server, we'd have to build a binary that runs the jobserver and spawns cargo with correct CARGO_MAKEFLAGS (during `./mach build).  But since make already has the jobserver functionality,  it *might* be simpler be to wrap the cargo invocation in a dummy GNU Make invocation in a way that forwards all arguments to cargo. But these approaches seem less than ideal to me as they add more moving parts to the build system.

I guess pushing for/contributing to upstream fix (either in SM or cargo adopting jobslot) and keeping make pinned via nix seems better,  now that we have the ability to use nix on other distros with your recent PR!

**Comment by redfire75369** [comment]
I was doing some investigation on this and I found that `cargo` is currently not planning for a named pipe-based make job server (also not implemented in `jobserver`). See [rust-lang/cargo#13843] due to incompatibility.

I was messing around with `jobslot`, and the code is quite short and simple.

<details>
<summary>Basic Jobserver</summary>

```rs
use std::error::Error;
use std::num::NonZeroUsize;
use std::process::Command;
use std::thread::available_parallelism;

use lexopt::{Arg, Parser, ValueExt};
use jobslot::Client;

fn main() -> Result<(), Box<dyn Error>> {
        let mut command = None;
        let mut jobs = available_parallelism().map_or(4, NonZeroUsize::get);

        let mut parser = Parser::from_env();
        while let Some(arg) = parser.next()? {
                match arg {
                        Arg::Short('j') | Arg::Long("jobs") => {
                                jobs = parser.value()?.parse()?;
                        }
                        Arg::Short(_) | Arg::Long(_) => {
                                println!("Usage: build-server [-j|--jobs=JOBS] <command>");
                                return Ok(());
                        }
                        Arg::Value(cmd) => {
                                command = Some(cmd);
                        }
                }

                if command.is_some() {
                        break;
                }
        }

        let Some(command) = command else {
                println!("Usage: build-server [-j|--jobs=JOBS] <command>");
                return Ok(());
        };

        let mut cmd = Command::new(command);

        for arg in parser.raw_args()? {
                cmd.arg(arg);
        }

        let client = Client::new_with_fifo(jobs)?;
        let status = client.configure_make_and_run_with_fifo(&mut cmd, |cmd| cmd.status())?;

        if status.success() {
                Ok(())
        } else {
                Err("Error occurred while running command".into())
        }
}
```

</details>
